src/plugins/crypto/gpg.c

```diff
@@ -0,0 +1,289 @@
+ *
+ * @param conf holds the backend/plugin configuration
+ * @param errorKey holds the error description in case of failure
+ * @param msgKey holds the master password to be decrypted. Note that the content of this key will be modified.
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoGpgDecryptMasterPassword (KeySet * conf, Key * errorKey, Key * msgKey)
+{
+	if (inTestMode (conf))
+	{
+		char * argv[] = { "", "--trust-model", "always", "-d", NULL };
+		return elektraCryptoGpgCall (conf, errorKey, msgKey, argv, 5);
+	}
+	else
+	{
+		char * argv[] = { "", "-d", NULL };
+		return elektraCryptoGpgCall (conf, errorKey, msgKey, argv, 3);
+	}
+}
+
+/**
+ * @brief call the gpg binary to perform the requested operation.
+ *
+ * @param conf holds the backend/plugin configuration
+ * @param errorKey holds the error description in case of failure
+ * @param msgKey holds the message to be transformed
+ * @param argv array holds the arguments passed on to the gpg process
+ * @param argc contains the number of elements in argv, i.e. the size of argv
+ *
+ * @retval 1 on success
+ * @retval -1 on failure
+ */
+int elektraCryptoGpgCall (KeySet * conf, Key * errorKey, Key * msgKey, char * argv[], size_t argc)
+{
+	pid_t pid;
+	int status;
+	int pipe_stdin[2];
+	int pipe_stdout[2];
+	kdb_octet_t * buffer = NULL;
+	const ssize_t bufferSize = 2 * keyGetValueSize (msgKey);
+	ssize_t outputLen;
+
+	assert (argc > 1);
+
+	// initialize pipes
+	if (pipe (pipe_stdin))
+	{
+		ELEKTRA_SET_ERROR (ELEKTRA_ERROR_CRYPTO_GPG_FAULT, errorKey, "Pipe initialization failed");
+		return -1;
+	}
+
+	if (pipe (pipe_stdout))
+	{
+		ELEKTRA_SET_ERROR (ELEKTRA_ERROR_CRYPTO_GPG_FAULT, errorKey, "Pipe initialization failed");
+		closePipe (pipe_stdin);
+		return -1;
+	}
+
+	// allocate buffer for gpg output
+	// estimated maximum output size = 2 * input (including headers, etc.)
+	if (!(buffer = elektraMalloc (bufferSize)))
+	{
+		ELEKTRA_SET_ERROR (ELEKTRA_ERROR_CRYPTO_GPG_FAULT, errorKey, "Memory allocation failed");
+		closePipe (pipe_stdin);
+		closePipe (pipe_stdout);
+		return -1;
+	}
+
+	// sanitize the argument vector
+	argv[0] = getGpgBinary (conf);
+	argv[argc - 1] = NULL;
+
+	// fork into the gpg binary
+	switch (pid = fork ())
+	{
+	case -1:
+		// fork() failed
+		ELEKTRA_SET_ERROR (ELEKTRA_ERROR_CRYPTO_GPG_FAULT, errorKey, "fork failed");
+		closePipe (pipe_stdin);
+		closePipe (pipe_stdout);
+		elektraFree (buffer);
+		return -1;
+
+	case 0:
+		// start of the forked child process
+		close (pipe_stdin[1]);
+		close (pipe_stdout[0]);
+
+		// redirect stdin to pipe
+		close (STDIN_FILENO);
+		dup (pipe_stdin[0]);
+		close (pipe_stdin[0]);
+
+		// redirect stdout to pipe
+		close (STDOUT_FILENO);
+		dup (pipe_stdout[1]);
+		close (pipe_stdout[1]);
+
+		// finally call the gpg executable
+		if (execv (argv[0], argv) < 0)
+		{
+			ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_GPG_FAULT, errorKey, "failed to start the gpg binary: %s", argv[0]);
```

Comment:
Can you improve this error?
E.g. check if the gpg binary exists and is executable.